### ISSUE TITLE
Track custom events being `$dispatch`-ed

### ISSUE THREAD

**Issue by HugoDF** [Te7a-Houdini/alpinejs-devtools#48]
This is again inspired by Vue.js Devtool's "events" tab, obviously Alpine.js doesn't use synthetic events or emit events to communicate between components.

Instead we should register a top-level listener for events being dispatched with `$dispatch` (which we can detect) eg. we can read `@click="$dispatch('my-custom-event')` which means we should listen to `'my-custom-event'` at the top-level (using document.addEventListener) and then report their contents in a tab.

Any thoughts/ideas limitations you can think of @ryangjchandler @Te7a-Houdini ?

**Comment by HugoDF** [comment]
Here's the relevant code in Alpine, if we can hook into `CustomEvent` or `el.dispatchEvent` that could work

https://github.com/alpinejs/alpine/blob/3c3e8b801a7b00a982dd54290cf656f051e915d9/src/component.js#L369

**Comment by ryangjchandler** [comment]
When you say hook in, do you mean monkey-patching the prototype?

**Comment by HugoDF** [comment]
> When you say hook in, do you mean monkey-patching the prototype?

Pretty much hahaha

**Comment by ryangjchandler** [comment]
> > When you say hook in, do you mean monkey-patching the prototype?
> 
> Pretty much hahaha

Yeah, I think that could work. Would just be a case of extending `EventTarget.dispatchEvent` and checking that the element in question has a parent or itself with `__x` property?

**Comment by ryangjchandler** [comment]
I think we should do the page -> devtools communication refactor first.

**Comment by HugoDF** [comment]
It turns out that it's possible to overwrite a magic property 👀 , so we can do inject some code that will replace `$dispatch` and send debug info on call 👍 

https://codepen.io/hugodf/pen/xxOvqpg

Other interesting note is that `$dispatch` basically doesn't need to be in core any more 😂

**Comment by ryangjchandler** [comment]
@HugoDF Legendary, hadn't even thought about that.

**Comment by KevinBatdorf** [comment]
> It turns out that it's possible to overwrite a magic property 👀

Oh nice! We can implement both `$refs` and `$dispatch` both now :D

**Comment by HugoDF** [comment]
> > It turns out that it's possible to overwrite a magic property 👀
> 
> Oh nice! We can implement both `$refs` and `$dispatch` both now :D

well see this update

> Overwriting $dispatch doesn't behave the same as Alpine implementation see https://codepen.io/hugodf/pen/GRjKgNO?editors=1111 the overriden $event.target is not correct (it's the root of the component, not the element from which $dispatch is called).
